### ISSUE TITLE
**InvalidArgumentError**:  Incompatible shapes: [2] vs. [8]

### ISSUE THREAD

**Issue by reyhan1994** [leondgarse/Keras_insightface#32]
I encounter the error(bath size Multiplied by 4) when I try to train but when I put batch_size=1 there is no error.
basic_model = models.buildin_models("mobilenet", dropout=0, emb_shape=512, output_layer='E')
tt = Train(data_path, save_path='TT_mobilenet_base_emb512_dr0_bs400.h5', eval_paths=eval_paths,
    basic_model=basic_model, model=None, lr_base=0.1, lr_decay=0.1, lr_decay_steps=[20, 30],
    batch_size=2, random_status=0)

**InvalidArgumentError**:  Incompatible shapes: [2] vs. [8]
	 [[node Equal (defined at <ipython-input-20-8757b735c9ce>:316) ]] [Op:__inference_train_function_20481]

Function call stack:
train_function

**Comment by leondgarse** [comment]
I didn't meet that issue, the place where `batch_size` may multiply with `4` is the `triplet` case, but it's `batch_size = batch_size // image_per_class * image_per_class`. Can you share the full log?

**Comment by reyhan1994** [comment]
/thanks for your reply. I solved it. was related to the size of my inputs. my other question is about to overfit or underfit, how can I check them? otherwise, how can I achieve validation loss?
because I train my own dataset without lfw format(.bin), I don't use evals.py and eval callbacks and I just test with eval_folder

**Comment by leondgarse** [comment]
I once thought about that, but it's not a common case. As the common used `lfw` / `agedb_30` is created for that job, so we can compare with others implementation. The reason why they are common is that their test cases are more general.

You may create your own `bin` file that defines which two images are same, which are not. I think is a better way using your own evaluating data in training process.
Another way is define your own `eval_callback` replacing the `__eval_func__` part in `evals.py`.

**Comment by reyhan1994** [comment]
thanks for your reply. I tried it and it needs to generate pairs.txt, how can I be sure that I generate it correctly? can I calculate issame_list myself to check the result of the pair.txt?(for a small dataset)

**Comment by leondgarse** [comment]
I' m not familiar with how the `pair.txt` is generated, but the final `bin` file just need two part, [image_data] and [is_same_list]
If we have folder:
```sh
.               # dataset folder
├── 1           # person folder
│   ├── 10.jpg # face image
│   └── 11.jpg # face image
├── 2           # person folder
│   ├── 20.jpg # face image
│   └── 21.jpg # face image
```
Then the `bin` file can be:
```py
image_data = [10_data, 11_data, 20_data, 21_data, 10_data, 20_data, 10_data, 21_data, 11_data, 20_data, 11_data, 21_data]
is_same_list = [1, 1, 0, 0, 0, 0]
```
And yes, you may check the `is_same_list`, or it will be too large. It maybe a random pick, or calculating the distance and pick the hard ones.

**Comment by reyhan1994** [comment]
thanks a lot, which accuracy do you recommend (PCA=False or True)? and what about calculating tpr or far?

**Comment by leondgarse** [comment]
The `PCA accuracy` is actually calculating the `tpr` and `far`, then calculates accuracy according to them.
```py
acc = float(tp + tn) / dist.size
```
My `eval_callback` is calculating accuracy by pick each thresh value in `far ∈ [0 ~ 0.1]`, and pick the max accuracy value.
```py
t_steps = int(0.1 * ff.shape[0])
acc_count = np.array([(tt > vv).sum() + (ff <= vv).sum() for vv in ff[-t_steps:]])
acc_max_indx = np.argmax(acc_count)
```
If set `PCA_acc=True`, then both these two accuracy values is calculated, but they are independent.
My recommendation is... `False`, because it's faster. But still it's your choice.
